pkg/api/merge.go

```diff
@@ -0,0 +1,368 @@
+	}
+	return partial
+}
+
+func mergeResourcePurchasePlan(complete, partial *ResourcePurchasePlan) *ResourcePurchasePlan {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &ResourcePurchasePlan{}
+	case partial == nil:
+		partial = &ResourcePurchasePlan{}
+	}
+
+	return &ResourcePurchasePlan{
+		Name:          mergeString(complete.Name, partial.Name),
+		Product:       mergeString(complete.Product, partial.Product),
+		PromotionCode: mergeString(complete.PromotionCode, partial.PromotionCode),
+		Publisher:     mergeString(complete.Publisher, partial.Publisher),
+	}
+}
+
+func mergeTags(complete, partial map[string]string) map[string]string {
+	if complete == nil && partial == nil {
+		return nil
+	}
+
+	// TODO: is this the semantic that we want for mergeTags?
+
+	rv := make(map[string]string, len(complete))
+	for k, v := range complete {
+		rv[k] = v
+	}
+	for k, v := range partial {
+		rv[k] = v
+	}
+	return rv
+}
+
+func mergeProperties(complete, partial *Properties) *Properties {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &Properties{}
+	case partial == nil:
+		partial = &Properties{}
+	}
+
+	return &Properties{
+		ProvisioningState:       mergeProvisioningState(complete.ProvisioningState, partial.ProvisioningState),
+		OpenShiftVersion:        mergeString(complete.OpenShiftVersion, partial.OpenShiftVersion),
+		PublicHostname:          mergeString(complete.PublicHostname, partial.PublicHostname),
+		FQDN:                    mergeString(complete.FQDN, partial.FQDN),
+		RouterProfiles:          mergeRouterProfiles(complete.RouterProfiles, partial.RouterProfiles),
+		AgentPoolProfiles:       mergeAgentPoolProfiles(complete.AgentPoolProfiles, partial.AgentPoolProfiles),
+		AuthProfile:             mergeAuthProfile(complete.AuthProfile, partial.AuthProfile),
+		ServicePrincipalProfile: mergeServicePrincipalProfile(complete.ServicePrincipalProfile, partial.ServicePrincipalProfile),
+		AzProfile:               mergeAzProfile(complete.AzProfile, partial.AzProfile),
+	}
+}
+
+func mergeProvisioningState(complete, partial ProvisioningState) ProvisioningState {
+	if partial == "" {
+		return complete
+	}
+	return partial
+}
+
+func mergeRouterProfiles(complete, partial []RouterProfile) []RouterProfile {
+	if complete == nil && partial == nil {
+		return nil
+	}
+
+	// preconditions: each RouterProfile in partial is named uniquely; each
+	// RouterProfile in complete is named uniquely
+
+	partialMap := make(map[string]RouterProfile, len(partial))
+	for _, p := range partial {
+		partialMap[p.Name] = p
+	}
+
+	rv := make([]RouterProfile, 0, len(complete))
+
+	for _, c := range complete {
+		if p, found := partialMap[c.Name]; found {
+			rv = append(rv, *mergeRouterProfile(&c, &p))
+			delete(partialMap, c.Name)
+		} else {
+			rv = append(rv, c)
+		}
+	}
+
+	for _, p := range partial {
+		if _, found := partialMap[p.Name]; !found {
```

Comment:
Trying to work through this logic and it seems that partial still contains all the entries which may have been deleted from `partialMap` so I'd expect this is appending things it shouldn't.  For instance:

```
# after partialMap loop
partial := ["foo"]
complete := ["foo"]
partialMap := {"foo": {//partialFoo}}
rv := []

# after complete loop
partial := ["foo"]
complete := ["foo"]
partialMap := {} 
rv := [mergedFoo]

# after this loop
partial := ["foo"]
complete := ["foo"]
partialMap := {} 
rv := [foo (mergedFoo), foo (partialFoo)]

Reply:
Still looking, but I don't think so.
In the complete loop, a p is deleted from partialMap whenever it corresponds to a c.
In the following loop, ps are added in order from partial only when they still exist in partialMap (i.e. did not correspond to a complete, were not previously added).

Reply:
The key piece here `if _, found := partialMap[p.Name]; !found` at the end

Reply:
```go
func TestMergeRouterProfile(t *testing.T) {
	complete := []RouterProfile{RouterProfile{Name: "foo"}}
	partial := []RouterProfile{RouterProfile{Name: "foo"}}
	merged := mergeRouterProfiles(complete, partial)
	if len(merged) == 2 {
		t.Errorf("too many: %v", merged)
	}
}
```
```
GOROOT=/home/pweil/.gimme/versions/go1.9.linux.amd64 #gosetup
GOPATH=/home/pweil/codebase/openshift-azure #gosetup
/home/pweil/.gimme/versions/go1.9.linux.amd64/bin/go test -c -i -o /tmp/___TestMergeRouterProfile_in_github_com_openshift_openshift_azure_pkg_api github.com/openshift/openshift-azure/pkg/api #gosetup
/tmp/___TestMergeRouterProfile_in_github_com_openshift_openshift_azure_pkg_api -test.v -test.run ^TestMergeRouterProfile$ #gosetup
	merge_test.go:10: too many: [{foo  } {foo  }]

Process finished with exit code 1
```

Reply:
Boo :(  Should be `found`, not `!found`.